UI Concepts pt. 1 (#1253)
square/workflow: docs/userguide/ui-concepts.md

# Workflow UI

This page provides a high level overview of Workflow UI, the companion that allows [Workflow Core](../core-workflow) to drive Android and iOS apps.

## Separation of Concerns

Workflow maintains a rigid separation between its core runtime and its UI support.
The [Workflow Core](../core-workflow) modules are strictly Swift and Kotlin, with no dependencies on any UI framework.
Dependencies on Android and iOS are restricted to the Workflow UI modules, as you would expect.
This innate separation naturally puts developers on a path to avoid entangling view concerns with their app logic.

And note that we say "app logic" rather than "business logic."
In any interesting app, the code that manages navigation and other UI-releated behavior is likely to dwarf that for what we typically think of as model concerns, in both size and complexity.

> We're all pretty good at capturing business concerns in tidy OO models of items for sale, shopping carts, payment cards and the like, nicely decoupled from the UI world.

Reviewer: When I read this, I felt that 'OO' kind of interrupted the flow.

```suggestion
We're all pretty good at capturing business concerns in tidy object-oriented models of items for sale, shopping carts, payment cards and the like, nicely decoupled from the UI world.
```

Reviewer 2: Same, good change I think.

Author: Nice, will do this in the next update.

# Workflow UI

This page provides a high level overview of Workflow UI, the companion that allows [Workflow Core](../core-workflow) to drive Android and iOS apps.

## Separation of Concerns

Workflow maintains a rigid separation between its core runtime and its UI support.
The [Workflow Core](../core-workflow) modules are strictly Swift and Kotlin, with no dependencies on any UI framework.

> Those dependencies are restricted to the Workflow UI modules, as you would expect, with Swift UI relying on iOS, and Kotlin UI on Android.

Reviewer 3: `Swift UI` could be mistaken for `SwiftUI` — Apple's native declarative UI framework (similar to Compose, or Blueprint).

Author: Yikes, thanks.

Author: ```
Dependencies on Android and iOS are restricted to the Workflow UI modules, as you would expect.
```

This innate separation naturally puts developers on a path to avoid entangling view concerns with their app logic.

!!! info "Separate both ways"

>     This strictness goes in both directions: neither UI library depends upon either runtime. They could easily be driven by any other stream of view models, and in view system tests they often are.

Reviewer 4: I'm not sure if we've honored this on the Swift side. We have an explicit [dependency](https://github.com/square/workflow-swift/blob/main/WorkflowUI.podspec#L21) on `Workflow`.

Author: Oh, that's too bad. Any idea why?

Reviewer 2: Take this with a grain of salt as I'm still very new to the code base, but the dependency was introduced in 2de5e98c2 and it looks like it's primarily because ContainerViewController imports Workflow and references WorkflowHost, AnyWorkflowConvertible, etc.

Author: Thanks for the spelunking. I just deleted this block, it didn't seem important.

    The "rendering" terminology does show up in the UI API and docs, but strictly as a convention, to make it easier to understand where [Workflow Core](../core-workflow) and Workflow UI are designed to bolt together.

And note that we say "app logic" rather than "business logic."
In any interesting app, the code that manages navigation and other UI-releated behavior is likely to dwarf that for what we typically think of as model concerns, in both size and complexity. 

We're all pretty good at capturing business concerns in tidy OO models of items for sale, shopping carts, payment cards and the like, nicely decoupled from the UI world.
But the rest of the app, and in particular the bits about how our users navigate it? Traditionally it's hard to keep that app logic code centralized, so that you can see what's going on; and even harder to keep it decoupled from your view system, so that it's easy to test.
The strict divide between Workflow UI and the runtime makes it easy.

## What's a Screen?

> Most Workflow implementations produce pure-Swift / pure-Kotlin renderings that can serve as view models.

Reviewer 3: What's a Rendering? Like, it'll be awesome to see how rendering and screen relate to each other, and what is the right way to call things depending on context.

Author: How about if I link to the glossary entry for that? Aaand of course there are no automatic bookmarks for DL entries, yay.

https://square.github.io/workflow/glossary/#:~:text=is%20strongly%20discouraged.-,Rendering,-The%20externally%20available

Author: Also note that rendering is already described in some depth in the core docs. Maybe should link to that instead? 

https://square.github.io/workflow/userguide/core-workflow/#workflows-produce-an-external-representation-of-their-state-via-rendering

Reviewer 3: Sounds good, and we could explain that Screen could be a derivative of Rendering, tailored to be responsible for UI representation and interaction (while Rendering itself is kind of a public snapshot of the state; but snapshot, and not The Snapshot). Damn, workflow terminology already employed every possible term I could think of :)

Author: I'm wary of going too far into the weeds here.

Such a rendering provides enough data to paint a UI, and functions to be called in response to UI events.

> The core service provided by Workflow UI is to transform these renderings into platform-specific UI structures, and to keep those views updated as new renderings are emitted.

Author: I was quietly inconsistent on purpose, trying to imply that UI structures are obviously lower-case-V views. Too distracting?

> The core service provided by Workflow UI is to transform these renderings into platform-specific UI structures, and to keep those views updated as new renderings are emitted.

Author: Too distracting.

```
The core service provided by Workflow UI is to transform `Screen` types into platform-specific view objects, and to keep those views updated as new `Screen` renderings are emitted.
```

The `Screen` [protocol](https://github.com/square/workflow-swift/blob/main/WorkflowUI/Sources/Screen/Screen.swift) / [interface](https://github.com/square/workflow-kotlin/blob/main/workflow-ui/core-common/src/main/java/com/squareup/workflow1/ui/Screen.kt) is the lynch pin that ties the two worlds together, and the basic UI building block for Workflow-driven apps.

> A screen is an object that can be presented as a basic 2D UI box, like an `android.view.View` or an `NSViewController`. And Workflow UI provides the glue that allows you to declare (at compile time!) that instances of `FooScreen : Screen` are used to drive `FooViewController`, `layout/foo_screen.xml`, or `@Composable fun Content(FooScreen, ViewEnvironment)`.

Author: Oops, `UIViewController`, my NeXTstep is showing.

=== "iOS"
    ```Swift
    struct WelcomeScreen: Screen {
        var name: String
        var onNameChanged: (String) -> Void
        var onLoginTapped: () -> Void

        func viewControllerDescription(environment: ViewEnvironment) -> ViewControllerDescription {
            return WelcomeViewController.description(for: self, environment: environment)
        }
    }

    private final class WelcomeViewController: ScreenViewController<WelcomeScreen> {
        // ...
    ```

=== "Android Classic"
    ```Kotlin
    data class HelloScreen(
      val message: String,
      val onClick: () -> Unit
    ) : AndroidScreen<HelloScreen> {
      override val viewFactory: ScreenViewFactory<HelloScreen> =
        fromViewBinding(HelloViewBinding::inflate) { r, _ ->
          helloMessage.text = r.message
          helloMessage.setOnClickListener { r.onClick() }
        }
    }
    ```

=== "Android Compose"
    ```Kotlin
    data class HelloScreen(
      val message: String,
      val onClick: () -> Unit
    ) : AndroidScreen<HelloScreen> by ComposeScreen {
      Button(onClick = onClick)

>       Text("message")

Author: Thanks

    }
    ```

After all the chest-thumping above about [Separation of Concerns](#separation-of-concerns), these code samples probably don't look very separate. That's because, while the libraries themselves are completely decoupled, they don't force that strict rigor on your app code.

If you aren't building, say, a core workflow module that you want to ship separately from its Android app, and an alternative command line interface, you'd probably gain nothing from enforced separation but boilerplate and runtime errors.

> And in practice, your Workflow unit tests won't call `viewControllerDescription()` or `viewFactory` and will build and run just fine against the JVM.

Author: ```suggestion
And in practice, your Workflow unit tests won't call `viewFactory` and will build and run just fine against the JVM. Likewise, at this point we've been building apps this way for hundreds of engineering years, and so far no one has called `viewControllerDescription()` and stashed a `UIViewController` in their workflow state. (This is not a challenge.)
```

> If you're one of the few who really and trully do need impermeable boundaries between your core and UI modules, that's not hard to get. Your `Screen` implementations can be defined completely separately from their view code and bound later.

Author: ```suggestion
If you're one of the few who really and truly do need impermeable boundaries between your core and UI modules, they aren't hard to get. Your `Screen` implementations can be defined completely separately from their view code and bound later.
```

=== "iOS"
    ```Swift
    // WelcomeScreen.swift
    struct WelcomeScreen {
        var name: String
        var onNameChanged: (String) -> Void
        var onLoginTapped: () -> Void
    }

    // WelcomeViewController.swift
    extension WelcomeScreen: Screen {
        func viewControllerDescription(environment: ViewEnvironment) -> ViewControllerDescription {
            return WelcomeViewController.description(for: self, environment: environment)
        }
    }

    private final class WelcomeViewController: ScreenViewController<WelcomeScreen> {
        // ...

>     ```

Reviewer 5: You can declare the struct with it implementing Screen and provide the implementation in an extension but it would have to be in the same swift module.

    data class HelloScreen(
      val message: String,
      val onClick: () -> Unit
    ) : Screen

    // HelloScreenViewFactory.kt
    object HelloScreenViewFactory: ScreenViewFactory<HelloScreen>
    by ScreenViewFactory.fromViewBinding(HelloViewBinding::inflate) { r, _ ->
          helloMessage.text = r.message
          helloMessage.setOnClickListener { r.onClick() }
        }
    }

    // HelloActivity.kt
    class HelloWorkflowActivity : AppCompatActivity() {
      override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        val model: HelloViewModel by viewModels()
        setContentView(
          WorkflowLayout(this).apply { take(lifecycle, model.renderings) }
        )
      }
    }

    class HelloViewModel(savedState: SavedStateHandle) : ViewModel() {
      private val viewRegistry = ViewRegistry(HelloScreenViewFactory)

      val renderings: StateFlow<Screen> by lazy {
        renderWorkflowIn(
          workflow = HelloWorkflow.mapRenderings { it.withRegistry(viewRegistry) },
          scope = viewModelScope,
          savedStateHandle = savedState
        )
      }
    }
    ```

!!! faq "Why \"Screen\"?"

>     We chose the name "Screen" because "View" would invite confusion with the like-named Android and iOS classes, and because "Box" didn't occur to us. 

Reviewer 2: Is it worth calling out that this is also unrelated to UIScreen, which is UIKit's representation of a hardware display?

Author: Missed this before I merged. Maybe add this?

> (No one seems to have been confused by the fact that `Screen` and iOS's `UIScreen` are unrelated.)

Reviewer 2: Yeah, I think so. Thanks!
